openoa/plant.py

```diff
@@ -1378,6 +1655,148 @@ def update_column_names(self, to_original: bool = False) -> None:
+        """
+        if self.scada is None:
+            raise AttributeError("This method can't be used unless `scada` data is provided.")
+        return self.scada.xs(turbine_id, level=1)
+
+    @property
+    def tower_id(self) -> np.ndarray:
+        """The 1D array of met tower IDs. This is created from the `asset` data, or unique IDs from the
+        tower data, if `asset` is undefined.
+        """
+        if self.asset is None:
+            return self.tower.index.get_level_values("id").unique()
+        return self.asset.loc[self.asset["type"] == "tower"].index.values
+
+    def tower_df(self, tower_id: str) -> pd.DataFrame:
+        """Filters `tower` on a single `tower_id` and returns the filtered data frame.
+
+        Args:
+            tower_id (str): The ID of the met tower to retrieve its data.
+
+        Returns:
+            pd.DataFrame: The met tower-specific data frame.
+        """
+        if self.tower is None:
+            raise AttributeError("This method can't be used unless `tower` data is provided.")
+        return self.tower.xs(tower_id, level=1)
+
+    @property
+    def asset_id(self) -> np.ndarray:
+        """The ID array of turbine and met tower IDs. This is created from the `asset` data, or unique
+        IDs from both the SCADA data and tower data, if `asset` is undefined.
+        """
+        if self.asset is None:
+            return np.concatenate([self.turbine_id, self.tower_id])
+        return self.asset.index.values
+
+    # NOTE: v2 AssetData methods
+
+    @cached_property
+    def asset_distance_matrix(self) -> pd.DataFrame:
+        """Calculates the distance between all assets on the site with `np.inf` for the distance
+        between an asset and itself.
+        """
+        ix = self.asset.index.values
+        distance = (
+            pd.DataFrame(
+                [i, j, self.asset.loc[i, "geometry"].distance(self.asset.loc[j, "geometry"])]
+                for i, j in itertools.combinations(ix, 2)
+            )
+            .pivot(index=0, columns=1, values=2)
+            .reset_index()
+            .fillna(0)
+            .loc[ix[:-1], ix[1:]]
+        )
+
+        # Insert the first column and last row because the self-self combinations are not produced in the above
+        distance.insert(0, ix[0], 0.0)
+        distance.loc[ix[-1]] = 0
+
+        # Unset the index and columns property names
+        distance.index.name = None
+        distance.columns.name = None
+
+        # Maintain v2 compatibility of np.inf for the diagonal
+        distance = distance + distance.values.T - np.diag(np.diag(distance.values))
+        np.fill_diagonal(distance.values, np.inf)
+        return distance
+
+    def calculate_nearest_neighbor(
+        self, turbine_ids: list | np.ndarray = None, tower_ids: list | np.ndarray = None
+    ) -> None:
+        """Finds nearest turbine neighbors for turbines and nearest tower neighbors for met towers.
```

Comment:
This made me think the method was going to only find the nearest turbine for turbine assets and nearest tower for tower assets. But it looks like it finds both the nearest turbine and tower for all assets. If that's correct, maybe update the description.